clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -772,6 +779,14 @@ synchronized void handleFailedBatch(ProducerBatch batch, RuntimeException except
         }
     }
 
+    /**
+     * Returns {@code true} if the given {@link ProducerBatch} has the same producer ID but a different epoch than the
+     * {@link #producerIdAndEpoch cached producer ID and epoch}.
+     */
+    synchronized boolean isStaleBatch(ProducerBatch batch) {
```

Comment:
Does this method need to by `synchronized`?

Also seems it could be `private`?

Reply:
> Does this method need to by `synchronized`?

No, not at present.

> Also seems it could be `private`?

Yes, it could. I might even just inline it in `handleFailedBatch()` to simplify things.

Reply:
I inlined the 'is stale batch' logic into `handleFailedBatch()`, so these issues are no longer present.